src/AdoPat/PatClient.cs

```diff
@@ -0,0 +1,139 @@
+                    top: PageSize,
+                    continuationToken: pagedPatTokens.ContinuationToken,
+                    sortByOption: SortByOptions.DisplayDate,
+                    cancellationToken: cancellationToken)
+                .ConfigureAwait(false);
+
+                foreach (PatToken pat in pagedPatTokens.PatTokens)
+                {
+                    pats.Add(pat);
+                }
+            }
+            while (!string.IsNullOrEmpty(pagedPatTokens.ContinuationToken));
+
+            return pats;
+        }
+
+        /// <summary>
+        /// Creates a new PAT with a new 'valid to' date by replicating an existing one.
+        /// </summary>
+        /// <param name="patToken">An existing <see cref="PatToken"/>.</param>
+        /// <param name="validTo">The new expiration date for the regenerated token.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns>A <see cref="Task"/> returning a <see cref="PatToken"/>.</returns>
+        public async Task<PatToken> RegeneratePatAsync(
+            PatToken patToken,
+            DateTime validTo,
+            CancellationToken cancellationToken = default)
+        {
+            // Regeneration is an option for existing PATs in the Azure DevOps UI,
+            // but from the API's viewpoint it is an abstraction over 3 operations.
+            //
+            // 1. Understand the metadata of the PAT using a GET call.
+            // 2. Create a new PAT with the old PAT's metadata using a POST call.
+            // 3. Revoke the old PAT using a DELETE call.
+            //
+            // We skip the initial GET call because the given patToken should have
+            // its own metadata. It risks being out of date depending on where it
+            // was fetched from.
+            //
+            // For more info see:
+            // https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/manage-personal-access-tokens-via-api?view=azure-devops#q-how-can-i-regeneraterotate-pats-through-the-api-i-saw-that-option-in-the-ui-but-i-dont-see-a-similar-method-in-the-api
+            PatTokenCreateRequest patTokenCreateRequest = new PatTokenCreateRequest
+            {
+                DisplayName = patToken.DisplayName,
+                Scope = patToken.Scope,
+                ValidTo = validTo,
+                AllOrgs = AllOrgs,
+            };
+
+            var renewedPatToken = await this.CreatePatAsync(
+                patTokenCreateRequest,
+                cancellationToken: cancellationToken)
+            .ConfigureAwait(false);
+
+            // XXX: If we create a new PAT, but fail to revoke the old PAT, should
+            // we continue or should we try to undo the creation and bail? It's
+            // possible this could fail because the token has already been revoked.
+            // Is that an error we should allow?

```

Comment:
Can two PATs have same `DisplayName`? 
If yes, then there will be a new PAT and an old PAT (which failed to get revoked) in system which I think should be okay, although I think we need to let the user know.
Since you are generating a new PAT before revoking it, I'm assuming you can have the same `DisplayName` for multiple tokens?

Reply:
Yes, it is possible that two PATs could have the same `DisplayName`, so this could cause confusion. We could try to alleviate that by adding our own unique prefix/suffix into the `DisplayName`, like when it was generated or something, but IDK how helpful that would actually be.

I posed this question because it's a _possible_ error case, but I don't think it's especially likely. Right now I'm trending towards not creating complicated retry logic or graceful recovery logic in favor of surfacing loud error messages to the user if something does go wrong, but I don't know if everyone will agree with me on that. 😅

Reply:
@kyle-rader: Do you have an opinion here?